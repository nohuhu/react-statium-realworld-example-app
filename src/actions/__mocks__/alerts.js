export const displayAlert = jest.fn();
