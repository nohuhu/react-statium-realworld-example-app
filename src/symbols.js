export const modalKey = Symbol.for('modal');
export const alertsKey = Symbol.for('alerts');
