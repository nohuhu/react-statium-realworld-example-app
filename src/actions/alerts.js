import { alertsKey } from '../symbols.js';

let id = 0;

export const displayAlert = async ({ state, set }, params) => {
  await set({
    [alertsKey]: [
      ...state[alertsKey],
      { ...params, id: id++ },
    ],
  })
};
