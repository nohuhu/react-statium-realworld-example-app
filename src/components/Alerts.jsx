import React from 'react';
import { useStore } from 'statium';

import { alertsKey } from '../symbols.js';

// This action is only used internally by the Alert widget
const closeAlert = async ({ state, set }, id) => {
  const alerts = state[alertsKey];

  const idx = alerts.findIndex(alert => alert.id === id);

  if (idx > -1) {
    alerts.splice(idx, 1);
    await set({ [alertsKey]: alerts })
  }
};

export const Alert = props => {
  const { dispatch } = useStore();
  const {
    id,
    type = 'primary',
    dismissible = true,
    text,
    timeout = 5000,
  } = props;

  const cls = `alert alert-${type} ${dismissible ? 'alert-dismissible' : ''}`

  let children = [...React.Children.toArray(props.children), text];

  if (dismissible) {
    const onClick = () => dispatch(closeAlert, id);

    if (timeout) {
      setTimeout(onClick, timeout);
    }

    children = [...children, (
      <button key="close" type="button"
        className="close"
        aria-label="Close"
        onClick={onClick}>
        <span aria-hidden="true">&times;</span>
      </button>
    )];
  }

  return (
    <div className={cls} role="alert">
      {children}
    </div>
  );
}

export const Alerts = () => {
  const { state: { [alertsKey]: alerts } } = useStore();

  return (
    <>
      { alerts.map(({ id, ...alert }) =>
        <Alert key={id} id={id} {...alert} />)
      }
    </>
  );
};
