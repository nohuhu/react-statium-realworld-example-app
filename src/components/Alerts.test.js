import React from 'react';
import { mount } from 'test/enzyme.js'

import { alertsKey } from '../symbols.js';
import Notifications from './NotificationProvider.jsx';
import { Alerts } from './Alerts.jsx';
import { displayAlert } from '../actions/alerts.js';

const alerts = [{
  id: 0,
  type: 'error',
  text: 'foo bar',
  dismissible: false,
}, {
  id: 1,
  text: 'blerg bonz',
}, {
  id: 2,
  type: 'success',
  text: 'throbbe zong',
}];

describe("Alerts", () => {
  let tree;

  beforeEach(() => {
    tree = mount(
      <Notifications>
        <Alerts />
      </Notifications>
    );
  });

  describe("rendering", () => {
    it("should render empty alert list by default", () => {
      expect(tree.find('Alerts')).toMatchSnapshot();
    });

    it("should render alert widgets correctly", async () => {
      await tree.find('Store[tag="Notifications"]').instance().set({
        [alertsKey]: alerts,
      });

      tree.update();

      expect(tree.find('Alerts')).toMatchSnapshot();
    });
  });

  describe("behavior", () => {
    beforeEach(async () => {
      await tree.find('Store[tag="Notifications"]').instance().dispatch(
        displayAlert,
        {
          type: "success",
          text: "plugh krabbe",
          timeout: 50,
        }
      );

      tree.update();
    });

    it("should display alert when action is dispatched", async () => {
      expect(tree.find('Alert')).toMatchSnapshot();
    });

    it("should close dismissible alert on timeout", async () => {
      // Dismiss timeout is set to 50 ms
      await sleep(100);

      expect(tree.find('Alerts').contains('Alert')).toBe(false);
    });

    it("should close dismissible alert on button click", async () => {
      tree.find('Alert button').simulate('click');

      // Give the handler enough time to fire
      await sleep(10);

      expect(tree.find('Alerts').contains('Alert')).toBe(false);
    });
  });
});