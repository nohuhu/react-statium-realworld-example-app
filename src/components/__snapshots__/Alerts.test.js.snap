// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alerts behavior should display alert when action is dispatched 1`] = `
<Alert
  id={0}
  key="0"
  text="plugh krabbe"
  timeout={50}
  type="success"
>
  <div
    className="alert alert-success alert-dismissible"
    role="alert"
  >
    plugh krabbe
    <button
      aria-label="Close"
      className="close"
      key="close"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
  </div>
</Alert>
`;

exports[`Alerts rendering should render alert widgets correctly 1`] = `
<Alerts>
  <Alert
    dismissible={false}
    id={0}
    key="0"
    text="foo bar"
    type="error"
  >
    <div
      className="alert alert-error "
      role="alert"
    >
      foo bar
    </div>
  </Alert>
  <Alert
    id={1}
    key="1"
    text="blerg bonz"
  >
    <div
      className="alert alert-primary alert-dismissible"
      role="alert"
    >
      blerg bonz
      <button
        aria-label="Close"
        className="close"
        key="close"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
  </Alert>
  <Alert
    id={2}
    key="2"
    text="throbbe zong"
    type="success"
  >
    <div
      className="alert alert-success alert-dismissible"
      role="alert"
    >
      throbbe zong
      <button
        aria-label="Close"
        className="close"
        key="close"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
  </Alert>
</Alerts>
`;

exports[`Alerts rendering should render empty alert list by default 1`] = `<Alerts />`;
